Provide dependencies artifacts during artifact transform

ljacomet opened the pull request:
This is an initial version of a solution #7648.

However the current implementation is only tested for usage with transforms triggered through an `ArtifactView`.

lptr commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultArtifactTransforms.java:
```diff
@@ -35,8 +36,8 @@ public DefaultArtifactTransforms(
-    public VariantSelector variantSelector(AttributeContainerInternal consumerAttributes, boolean allowNoMatchingVariants) {
-        return new AttributeMatchingVariantSelector(consumerProvidedVariantFinder, schema, attributesFactory, consumerAttributes.asImmutable(), allowNoMatchingVariants);
+    public VariantSelector variantSelector(AttributeContainerInternal consumerAttributes, boolean allowNoMatchingVariants, ConfigurationInternal configuration) {
+        return new AttributeMatchingVariantSelector(consumerProvidedVariantFinder, schema, attributesFactory, consumerAttributes.asImmutable(), allowNoMatchingVariants, configuration);
```
I see that we are now pushing the `Configuration` object down to the transform across half the world. It would really clarify things if we could a) call the parameter/variable something that designates its role and purpose instead of repeating the type name, or even better b) pass down only a type that exposes only the necessary functionality, and give that type a good name.

ljacomet replied:
Half the world ... it is a very short path compared to others 😉 
But you are perfectly right about both the type and the name ... as soon as we know exactly what will be required to get the `Set<ResolvedArtifactResult>`

ljacomet commented:
PR updated:
* Fixed a number of test failures ... CI will tell me if I got them all
* Added a test that shows that when using transforms through attributes on configuration, the files made visible are the transformed ones

ljacomet commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformingAsyncArtifactListener.java:
```diff
@@ -61,7 +64,7 @@ public boolean includeFileDependencies() {
     @Override
     public void fileAvailable(File file) {
-        TransformationSubject initialSubject = TransformationSubject.initial(file);
+        TransformationSubject initialSubject = TransformationSubject.initial(file, configuration);
```
This is the entry point where we do not have a `ComponentArtifactIdentifier` available and thus results in having to support a `null` when running the transform and deciding if we can provide the dependencies.

lptr replied:
IIUC, this is a standalone `File` artifact that has no way to have any dependencies. `TransformationSubject.inital(File)` should not even need the configuration, or anything, because we have no chance to find anything in there.

If we passed around some new type like `DependencyArtitfactLookup` instead of the whole `Configuration` object, we could use a no-op singleton of that here.

ljacomet replied:
Done

lptr commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformWithDependenciesIntegrationTest.groovy:
```diff
@@ -0,0 +1,170 @@
+
+    def setup() {
+        // TODO LJA Investigate, this makes little sense
+        executer.expectDeprecationWarning()
```
Need to fix this before we can merge.

ljacomet replied:
Investigating

wolfs replied:
@lptr Is it later already? What is the problem here?

lptr replied:
This is some kind of locking issue apparently, you can see the warning printed.

<details><summary>Here's the stacktrace</summary>
<p>

```text
The configuration :app:compileClasspath was resolved without accessing the project in a safe manner.  This may happen when a configuration is resolved from a thread not managed by Gradle or from a different project.  See https://docs.gradle.org/5.1-20181111230000+0000/userguide/troubleshooting_dependency_resolution.html#configuration_resolution_constraints for more details. This behaviour has been deprecated and is scheduled to be removed in Gradle 6.0.
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveToStateOrLater(DefaultConfiguration.java:521)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$2100(DefaultConfiguration.java:127)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationResolvableDependencies.assertArtifactsResolved(DefaultConfiguration.java:1343)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationResolvableDependencies.access$3300(DefaultConfiguration.java:1232)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationResolvableDependencies$LenientResolutionResult.resolve(DefaultConfiguration.java:1358)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationResolvableDependencies$LenientResolutionResult.getAllComponents(DefaultConfiguration.java:1395)
	at org.gradle.api.internal.artifacts.transform.DefaultArtifactTransformDependencies.getDependencies(DefaultArtifactTransformDependencies.java:48)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:104)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:326)
	at org.codehaus.groovy.runtime.metaclass.MethodMetaProperty$GetBeanMethodMetaProperty.getProperty(MethodMetaProperty.java:76)
	at org.codehaus.groovy.runtime.callsite.GetEffectivePojoPropertySite.getProperty(GetEffectivePojoPropertySite.java:63)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callGetProperty(AbstractCallSite.java:298)
	at FileSizer.transform(/Users/lptr/Workspace/gradle/gradle/subprojects/dependency-management/build/tmp/test files/ArtifactTransformWithDependenciesIntegrationTest/transform_can_acces...ifactView/73wxy/build.gradle:48)
	at org.gradle.api.internal.artifacts.transform.DefaultTransformer.transform(DefaultTransformer.java:49)
	at org.gradle.api.internal.artifacts.transform.DefaultTransformerInvoker$TransformerExecution.execute(DefaultTransformerInvoker.java:173)
	at org.gradle.internal.execution.impl.steps.ExecuteStep.execute(ExecuteStep.java:49)
	at org.gradle.internal.execution.impl.steps.CancelExecutionStep.execute(CancelExecutionStep.java:34)
	at org.gradle.internal.execution.impl.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:69)
	at org.gradle.internal.execution.impl.steps.TimeoutStep.execute(TimeoutStep.java:49)
	at org.gradle.internal.execution.impl.steps.CatchExceptionStep.execute(CatchExceptionStep.java:33)
	at org.gradle.internal.execution.impl.steps.CreateOutputsStep.execute(CreateOutputsStep.java:50)
	at org.gradle.internal.execution.impl.steps.SnapshotOutputStep.execute(SnapshotOutputStep.java:43)
	at org.gradle.internal.execution.impl.steps.SnapshotOutputStep.execute(SnapshotOutputStep.java:29)
	at org.gradle.internal.execution.impl.steps.CacheStep.executeWithoutCache(CacheStep.java:134)
	at org.gradle.internal.execution.impl.steps.CacheStep.lambda$execute$3(CacheStep.java:83)
	at java.base/java.util.Optional.orElseGet(Optional.java:361)
	at org.gradle.internal.execution.impl.steps.CacheStep.execute(CacheStep.java:82)
	at org.gradle.internal.execution.impl.steps.CacheStep.execute(CacheStep.java:36)
	at org.gradle.internal.execution.impl.steps.PrepareCachingStep.execute(PrepareCachingStep.java:33)
	at org.gradle.internal.execution.impl.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:38)
	at org.gradle.internal.execution.impl.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:23)
	at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:96)
	at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.lambda$execute$1(SkipUpToDateStep.java:91)
	at java.base/java.util.Optional.orElseGet(Optional.java:361)
	at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:91)
	at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:36)
	at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:34)
	at org.gradle.api.internal.artifacts.transform.DefaultTransformerInvoker.lambda$invoke$0(DefaultTransformerInvoker.java:105)
	at org.gradle.api.internal.artifacts.transform.DefaultTransformerInvoker.fireTransformListeners(DefaultTransformerInvoker.java:128)
	at org.gradle.api.internal.artifacts.transform.DefaultTransformerInvoker.lambda$invoke$1(DefaultTransformerInvoker.java:102)
	at org.gradle.api.internal.artifacts.transform.DefaultTransformerExecutionHistoryRepository.lambda$withWorkspace$1(DefaultTransformerExecutionHistoryRepository.java:80)
	at org.gradle.api.internal.artifacts.transform.GradleUserHomeWorkspaceProvider.lambda$withWorkspace$0(GradleUserHomeWorkspaceProvider.java:74)
	at org.gradle.cache.internal.ProducerGuard$AdaptiveProducerGuard.guardByKey(ProducerGuard.java:97)
	at org.gradle.api.internal.artifacts.transform.GradleUserHomeWorkspaceProvider.withWorkspace(GradleUserHomeWorkspaceProvider.java:70)
	at org.gradle.api.internal.artifacts.transform.DefaultTransformerExecutionHistoryRepository.withWorkspace(DefaultTransformerExecutionHistoryRepository.java:75)
	at org.gradle.api.internal.artifacts.transform.DefaultTransformerInvoker.invoke(DefaultTransformerInvoker.java:101)
	at org.gradle.api.internal.artifacts.transform.TransformationStep.transform(TransformationStep.java:53)
	at org.gradle.api.internal.artifacts.transform.TransformationNode$InitialTransformationNode$InitialArtifactTransformationStepOperation.run(TransformationNode.java:172)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:301)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:293)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:175)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.artifacts.transform.TransformationNode$InitialTransformationNode.execute(TransformationNode.java:119)
	at org.gradle.api.internal.artifacts.transform.TransformationNodeExecutor.execute(TransformationNodeExecutor.java:37)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:345)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:333)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:326)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:312)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
	at org.gradle.execution.plan.DefaultPlanExecutor.process(DefaultPlanExecutor.java:74)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.executeWithServices(DefaultTaskExecutionGraph.java:175)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.execute(DefaultTaskExecutionGraph.java:151)
	at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:41)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:40)
	at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:24)
	at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:46)
	at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:49)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:40)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:33)
	at org.gradle.initialization.DefaultGradleLauncher$ExecuteTasks.run(DefaultGradleLauncher.java:383)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:301)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:293)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:175)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.initialization.DefaultGradleLauncher.runTasks(DefaultGradleLauncher.java:247)
	at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:159)
	at org.gradle.initialization.DefaultGradleLauncher.executeTasks(DefaultGradleLauncher.java:134)
	at org.gradle.internal.invocation.GradleBuildController$1.execute(GradleBuildController.java:58)
	at org.gradle.internal.invocation.GradleBuildController$1.execute(GradleBuildController.java:55)
	at org.gradle.internal.invocation.GradleBuildController$3.create(GradleBuildController.java:82)
	at org.gradle.internal.invocation.GradleBuildController$3.create(GradleBuildController.java:75)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:183)
	at org.gradle.internal.work.StopShieldingWorkerLeaseService.withLocks(StopShieldingWorkerLeaseService.java:40)
	at org.gradle.internal.invocation.GradleBuildController.doBuild(GradleBuildController.java:75)
	at org.gradle.internal.invocation.GradleBuildController.run(GradleBuildController.java:55)
	at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:31)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:58)
	at org.gradle.tooling.internal.provider.ValidatingBuildActionRunner.run(ValidatingBuildActionRunner.java:32)
	at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:39)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:49)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:44)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:315)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:305)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:175)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:101)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:44)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:49)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:46)
	at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:78)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:46)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:31)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:42)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:28)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:78)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:52)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:56)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:35)
	at org.gradle.tooling.internal.provider.SessionScopeBuildActionExecuter.execute(SessionScopeBuildActionExecuter.java:68)
	at org.gradle.tooling.internal.provider.SessionScopeBuildActionExecuter.execute(SessionScopeBuildActionExecuter.java:38)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:37)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:26)
	at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:43)
	at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:29)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:60)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:32)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:55)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:41)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:48)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:32)
	at org.gradle.integtests.fixtures.executer.InProcessGradleExecuter.executeBuild(InProcessGradleExecuter.java:323)
	at org.gradle.integtests.fixtures.executer.InProcessGradleExecuter.doRun(InProcessGradleExecuter.java:245)
	at org.gradle.integtests.fixtures.executer.InProcessGradleExecuter.doRun(InProcessGradleExecuter.java:147)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.run(AbstractGradleExecuter.java:1047)
	at org.gradle.integtests.fixtures.executer.AbstractDelegatingGradleExecuter.doRun(AbstractDelegatingGradleExecuter.java:29)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.run(AbstractGradleExecuter.java:1047)
	at org.gradle.integtests.fixtures.executer.GradleExecuter$run$2.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:119)
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.succeeds(AbstractIntegrationSpec.groovy:212)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSite.invoke(PogoMetaMethodSite.java:167)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:58)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:156)
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.run(AbstractIntegrationSpec.groovy:200)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSite.invoke(PogoMetaMethodSite.java:167)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:58)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:156)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:168)
	at org.gradle.integtests.resolve.transform.ArtifactTransformWithDependenciesIntegrationTest.$spock_feature_5_0(ArtifactTransformWithDependenciesIntegrationTest.groovy:128)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.spockframework.util.ReflectionUtil.invokeMethod(ReflectionUtil.java:200)
	at org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:113)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:484)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:467)
	at org.spockframework.runtime.BaseSpecRunner.runFeatureMethod(BaseSpecRunner.java:408)
	at org.spockframework.runtime.BaseSpecRunner.doRunIteration(BaseSpecRunner.java:322)
	at org.spockframework.runtime.BaseSpecRunner$6.invoke(BaseSpecRunner.java:306)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:98)
	at org.spockframework.runtime.extension.IMethodInvocation$proceed.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:115)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:119)
	at org.gradle.test.fixtures.file.CleanupTestDirectoryExtension$FailureCleanupInterceptor.intercept(CleanupTestDirectoryExtension.groovy:57)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:97)
	at org.spockframework.runtime.extension.builtin.AbstractRuleInterceptor$1.evaluate(AbstractRuleInterceptor.java:37)
	at org.gradle.test.fixtures.file.AbstractTestDirectoryProvider$TestDirectoryCleaningStatement.evaluate(AbstractTestDirectoryProvider.java:90)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.spockframework.runtime.extension.builtin.TestRuleInterceptor.intercept(TestRuleInterceptor.java:39)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:97)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:475)
	at org.spockframework.runtime.BaseSpecRunner.runIteration(BaseSpecRunner.java:285)
	at org.spockframework.runtime.BaseSpecRunner.initializeAndRunIteration(BaseSpecRunner.java:275)
	at org.spockframework.runtime.BaseSpecRunner.runSimpleFeature(BaseSpecRunner.java:266)
	at org.spockframework.runtime.BaseSpecRunner.doRunFeature(BaseSpecRunner.java:260)
	at org.spockframework.runtime.BaseSpecRunner$5.invoke(BaseSpecRunner.java:243)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:484)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:467)
	at org.spockframework.runtime.BaseSpecRunner.runFeature(BaseSpecRunner.java:235)
	at org.spockframework.runtime.BaseSpecRunner.runFeatures(BaseSpecRunner.java:185)
	at org.spockframework.runtime.BaseSpecRunner.doRunSpec(BaseSpecRunner.java:95)
	at org.spockframework.runtime.BaseSpecRunner$1.invoke(BaseSpecRunner.java:81)
	at org.spockframework.runtime.BaseSpecRunner.invokeRaw(BaseSpecRunner.java:484)
	at org.spockframework.runtime.BaseSpecRunner.invoke(BaseSpecRunner.java:467)
	at org.spockframework.runtime.BaseSpecRunner.runSpec(BaseSpecRunner.java:73)
	at org.spockframework.runtime.BaseSpecRunner.run(BaseSpecRunner.java:64)
	at org.spockframework.runtime.Sputnik.run(Sputnik.java:63)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

</p>
</details>

lptr replied:
My understanding is that @ljacomet will take a look next week unless I get to it before that.

wolfs replied:
Can you open a follow up issue for this?

lptr replied:
https://github.com/gradle/gradle/issues/7729

lptr commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformWithDependenciesIntegrationTest.groovy:
```diff
@@ -0,0 +1,170 @@
+
+        then:
+        output.count("Transforming") == 4
+        // The asserts below should not have the .txt part
```
Shall this be marked as `TODO`? Also, we should probably use `@ToBeImplemented` on this one.

ljacomet replied:
Done for the `@ToBeImplemented`

lptr commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformerInvoker.java:
```diff
@@ -203,25 +212,60 @@ public int hashCode() {
+        private final ResolvableDependencies incoming;
 
-        public TransformerExecution(File primaryInput, Transformer transformer, HashCode persistentCacheKey, File workspace) {
+        public TransformerExecution(File primaryInput, Transformer transformer, HashCode persistentCacheKey, File workspace, @Nullable ComponentArtifactIdentifier artifactId, ResolvableDependencies incoming) {
```
We don't use `incoming` for anything if `artifactId` is `null`, so we could as well skip both. I think it would be better to pass something that is basically a `Supplier<FileCollection>` -- for the no `artifactId` case it could return an empty collection always.

ljacomet commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/BuildDependenciesOnlyVisitedArtifactSet.java:
```diff
@@ -31,16 +32,18 @@
         this.unresolvedDependencies = unresolvedDependencies;
         this.artifactsResults = artifactsResults;
         this.artifactTransforms = artifactTransforms;
+        this.configuration = configuration;
```
My thinking is that this gets cleaned up once we get this feature working for transforms triggered by configuration level attributes.

lptr replied:
I've changed this to pass `ResolvableDependencies` instead.

ljacomet commented on subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java:
```diff
@@ -157,7 +158,11 @@ private void executeWithServices(ProjectExecutionServiceRegistry projectExecutio
-            buildOperationExecutor.run(new NotifyTaskGraphWhenReady(this, graphListeners.getSource(), gradleInternal));
+            ProjectStateRegistry projectStateRegistry = gradleInternal.getServices().get(ProjectStateRegistry.class);
+            // We know that we're running single-threaded here, so we can use lenient project locking
+            projectStateRegistry.withLenientState(() -> {
```
@ghale Is that what you had in mind?

ljacomet commented:
@lptr Well, sorry it took that long. Now that we are at perf tests, I do not expect failures ... but hey, who knows.
If you validate the PR, I'll merge if the build is all good.

ljacomet commented:
Well, looks like forcing dependency graph realization has a high cost. Not sure what's the next step there ...

lptr commented:
@ljacomet I see 0.17% and 0.03% regressions on the Android tests. These seem very much acceptable. Or am I missing something?

lptr commented:
OK, so it looks like the reports are faulty, and the regression is actually 17% and 3%. Back to the drawing board...

wolfs commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformWithDependenciesIntegrationTest.groovy:
```diff
@@ -0,0 +1,173 @@
+    apply plugin: 'java-library'
+
+    dependencies {
+        api 'org.slf4j:slf4j-api:1.7.25'
```
Can we add some external dependencies to `lib` which have transitive dependencies? So we see them provided to the transformation of the `lib` project?

wolfs commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformWithDependenciesIntegrationTest.groovy:
```diff
@@ -0,0 +1,173 @@
+"""
+    }
+
+    def "transform can access artifact dependencies as FileCollection when using ArtifactView"() {
```
We need a test for chained transformations as well. What dependencies will be provided to transformations further down the chain?

lptr replied:
It's always the dependencies of the initial artifact.

wolfs replied:
That is what I expected. Can we add a test?

lptr replied:
Yeah, so that's a problem. We'll need to address this. See main thread of conversation for details.

wolfs commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformationNodeFactory.java:
```diff
@@ -46,30 +47,30 @@ public boolean visitArtifactSet(ResolvedArtifactSet set) {
     }
 
-    private static List<TransformationStep> unpackTransformation(Transformation transformation) {
+    static List<TransformationStep> unpackTransformation(Transformation transformation) {
```
Shouldn't that stay private?

lptr replied:
It should, I'll change it back.

wolfs commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformer.java:
```diff
@@ -72,8 +74,14 @@ public DefaultTransformer(Class<? extends ArtifactTransform> implementationClass
-        return instantiator.newInstance(implementationClass, parameters.isolate());
+    private ArtifactTransform newTransformer(ArtifactTransformDependencies artifactTransformDependencies) {
+        DefaultServiceRegistry registry = new DefaultServiceRegistry();
+        registry.addProvider(new Object() {
```
If we use the `add(Class, T)` method, do we then not require scanning the methods of the anonymous inner class?

lptr replied:
Good point. We still scan the methods of the `DefaultServiceRegistry` instance, but we can take care of that separately.

wolfs commented on subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/ConsumerProvidedVariantFinderTest.groovy:
```diff
@@ -325,7 +331,7 @@ class ConsumerProvidedVariantFinderTest extends Specification {
     }
 
-    private static TransformationSubject initialSubject(String path) {
+    private TransformationSubject initialSubject(String path) {
```
Why is this method not static anymore?

lptr replied:
Turning it back to `static`.

wolfs commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformer.java:
```diff
@@ -72,8 +74,14 @@ public DefaultTransformer(Class<? extends ArtifactTransform> implementationClass
-    private ArtifactTransform newTransformer() {
-        return instantiator.newInstance(implementationClass, parameters.isolate());
+    private ArtifactTransform newTransformer(ArtifactTransformDependencies artifactTransformDependencies) {
+        DefaultServiceRegistry registry = new DefaultServiceRegistry();
```
If creating the service registry is expensive, should we create the registry only when the dependencies are actually required by the transformer?

lptr replied:
I'd change things around and make a cheap service registry instead. Let's address that in a separate PR.

lptr commented:
OK, so after a long discussion with @wolfs we discovered that we don't need to fully resolve configurations to provide artifact transforms with their dependencies—this fixed the performance regression, so all is well...

lptr commented:
...except we also figured out that exposing the _initial_ dependencies to each transform in a chain makes no sense, and we need to do more work to make this feature work.

Consider the following chain of first extracting artifacts then obfuscating their contents:

`[original] ---A--> [extracted] ---B--> [obfuscated]`

For the transformation `A` we expose dependencies in the `[original]` variant, which is what we want. However, we also expose dependencies to `B` as `[original]` instead of `[extracted]`.

Transforms should be _context insensitive,_ and should only care and know about their immediate _source_ and _target._ The current implementation breaks this assumption, because it makes the format of the exposed dependencies _depend on_ the chain the transformation takes place in.

Because we don't depend on transformed artifacts, we also miss the piece where we actually schedule transformations in a way that transformed dependencies become available before a transform is executed.

We need to fix both of these before we can merge this PR. :(

lptr commented:
@wolfs I addressed most of your feedback. Let's merge this PR and address the two remaining issues as separate ones:

- capture dependencies as inputs to the transformation
- expose dependencies of the transformation in terms of its source, not in terms of the initial variant

melix commented on subprojects/core-api/src/main/java/org/gradle/api/artifacts/transform/ArtifactTransformDependencies.java:
```diff
@@ -0,0 +1,47 @@
+     * Returns the dependency artifacts of the artifact being transformed.
+     * The order of the files match that of the dependencies in the source artifact view.
+     */
+    Iterable<File> getDependencies();
```
:x: can we call this `getArtifacts()`? It may be that we need to provide "real" dependencies in the future, with more metadata.

lptr replied:
I'd rather call this `getFiles()`, since we are returning files.

melix commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformWithDependenciesIntegrationTest.groovy:
```diff
@@ -0,0 +1,173 @@
+
+    ArtifactTransformDependencies artifactDependencies
+
+    @Inject
```
I feels really weird to me that we inject the artifact dependencies, but that the actual artifact being transformed is passed as an argument to the `transform` method. I know this is done because not all transforms need it, but just wanted to highlight the pattern is a bit strange.

lptr replied:
Do you have suggestions about how to do this differently? What other options do we have?

We need a declarative way that can be inspected by Gradle without actually executing the transform to decide whether or not this information is going to be used as input. That means we can't make it always available.

Another option is what we do with `IncrementalTaskInputs`, i.e. we inject it into the `@TaskAction` method. I'm not a fan of that approach either, because it leads to weird questions like "what order do you execute things when they have multiple `@TaskAction` methods?", or "what if there's no `@TaskAction` method?".

The only downside I see with injected services is that there is a discoverability problem, but that problem exists with the `@TaskAction`-injected approach as well, and can be solved in several ways (e.g. by added documentation, implement a common `UserInjectableService` interface etc.).

Basically, I see three options to expose this information to a transform:

* expose always via some baked in parameter of method
* inject via constructor parameter
* inject via `@TransformAction` method parameter

Do you know of other solutions we could use? Of the above I think constructor injection is the only one worth considering.

melix commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformationSubject.java:
```diff
@@ -36,27 +37,33 @@ public static TransformationSubject initial(File file) {
+    /**
+     * Gives access to the artifacts of the dependencies of the subject of the transformation
+     */
+    public abstract ArtifactTransformDependencies getDependencies();
```
:o: `getArtifactDependencies()`?

lptr replied:
We have a plan to get rid of `TransformationSubject` altogether. Let's implement that instead.

lptr commented:
I'm going to merge this as soon as we have a green `master`.
